src/operators/mod.rs

```diff
@@ -100,896 +71,225 @@ pub(crate) fn inspect_epoch(inspector: &TdPyCallable, epoch: &u64, item: &TdPyAn
-                None => value,
-                Some(acc) => {
-                    let updated_acc = with_traceback!(
-                        py,
-                        reducer.call1(py, (acc.clone_ref(py), value.clone_ref(py)))
-                    )
-                    .into();
-                    debug!("reduce for key={key:?}: reducer={reducer:?}(acc={acc:?}, value={value:?}) -> updated_acc={updated_acc:?}",);
-                    updated_acc
-                }
-            };
-
-            let should_emit_and_discard_acc: bool = with_traceback!(py, {
-                let should_emit_and_discard_acc_pybool: TdPyAny =
-                    is_complete.call1(py, (updated_acc.clone_ref(py),))?.into();
-                should_emit_and_discard_acc_pybool.extract(py).map_err(|_err| PyTypeError::new_err(format!("return value of `is_complete` in reduce operator must be a bool; got `{should_emit_and_discard_acc_pybool:?}` instead")))
-            });
-            debug!("reduce for key={key:?}: is_complete={is_complete:?}(updated_acc={updated_acc:?}) -> should_emit_and_discard_acc={should_emit_and_discard_acc:?}");
+}
 
-            if should_emit_and_discard_acc {
-                StatefulUnaryLogicReturn {
-                    updated_state: None,
-                    output: Some(updated_acc),
-                    activate_after: None,
-                }
-            } else {
-                StatefulUnaryLogicReturn {
-                    updated_state: Some(updated_acc),
-                    output: None,
-                    activate_after: None,
-                }
-            }
-        })
-    } else {
-        StatefulUnaryLogicReturn {
-            updated_state: acc,
-            output: None,
-            activate_after: None,
+impl ReduceLogic {
+    /// Returns a function that can deserialize the result of
+    /// [`Self::snapshot`].
+    pub(crate) fn builder(
+        reducer: TdPyCallable,
+        is_complete: TdPyCallable,
+    ) -> impl Fn(Option<StateBytes>) -> Self {
+        move |resume_acc_bytes| {
+            let acc = resume_acc_bytes.map(|resume_acc_bytes| resume_acc_bytes.de());
+            Python::with_gil(|py| Self {
+                reducer: reducer.clone_ref(py),
+                is_complete: is_complete.clone_ref(py),
+                acc,
+            })
         }
     }
 }
 
-/// Wrapper struct so we have named fields.
-///
-/// [`StatefulUnary::stateful_unary`]'s logic can do a lot each
-/// activation so this helps.
-pub(crate) struct StatefulUnaryLogicReturn<R, D, I: IntoIterator<Item = R>> {
-    pub(crate) updated_state: Option<D>,
-    pub(crate) output: I,
-    pub(crate) activate_after: Option<Duration>,
-}
-
-/// Extension trait for [`Stream`].
-// Based on the good work in
-// https://github.com/TimelyDataflow/timely-dataflow/blob/0d0d84885672d6369a78cd9aff7beb2048390d3b/timely/src/dataflow/operators/aggregation/state_machine.rs#L57
-pub(crate) trait StatefulUnary<S: Scope, V: ExchangeData> {
```

Comment:
All of the Timely operators just used for recovery I moved into `crate::recovery`. I think we'll just keep Bytewax operator definitions in this module.